kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
                 LOGGER.warn("Unexpected message writing to upstream: {}", msg, new IllegalStateException());
             }
             ctx.write(msg, promise);
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    private <T> CompletableFuture<T> switchBackToNettyIfRequired(ChannelHandlerContext ctx, CompletableFuture<T> future) {
+        CompletableFuture<T> nettyDrivenFuture = new CompletableFuture<>();
+        future.whenComplete((t, throwable) -> {
+            if (ctx.executor().inEventLoop()) {
+                complete(nettyDrivenFuture, t, throwable);
+            }
+            else {
+                ctx.executor().execute(() -> {
+                    complete(nettyDrivenFuture, t, throwable);
+                });
+            }
+        });
+        return nettyDrivenFuture;
+    }
+
+    private static <T> void complete(CompletableFuture<T> nettyDrivenFuture, T t, Throwable throwable) {
+        if (throwable != null) {
+            nettyDrivenFuture.completeExceptionally(throwable);
+        }
+        else {
+            nettyDrivenFuture.complete(t);
+        }
+    }
+
+    private <T extends FilterResult> CompletableFuture<T> handleDeferredStage(ChannelHandlerContext ctx, CompletableFuture<T> stage) {
+        if (!stage.isDone()) {
+            inboundChannel.config().setAutoRead(false);
```

Comment:
Not for now - we need metrics about how long we stop owing to awaiting the async work of filters.

Reply:
#520